tarsier/tag_utils.ts

```diff
@@ -458,67 +461,124 @@ function insertTags(
           (child as HTMLElement).tagName.toLowerCase(),
         )
       ) {
-        return getElementToInsertInto(child as HTMLElement, idSpan);
+        return getElementToInsertInto(child as HTMLElement);
       }
     }
 
     trimTextNodeStart(element);
-
-    // Base case
     return element;
   }
 
-  const idToTagMetadata: { [key: number]: TagMetadata } = {};
-  let idNum = 0;
-
-  for (let el of elementsToTag) {
-    idToTagMetadata[idNum] = {
-      xpath: getElementXPath(el),
-    };
+  function getOpeningTag(el: HTMLElement): string {
+    return el.outerHTML.split(">")[0] + ">";
```

Comment:
this is unsafe. we should always avoid doing string manipulation if possible

for example,
```
<div id=">12" className="hi" data-attribute="something">...
``` 
would result in an opening tag of 
```
<div id=">
```

You can use this instead. Read up on it to see how it works/see if i'd be a good fit
```
document.querySelector('div').cloneNode().outerHTML
```